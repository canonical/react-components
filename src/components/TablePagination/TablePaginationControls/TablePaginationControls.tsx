import Button, { ButtonProps } from "components/Button";
import Icon from "components/Icon";
import Input from "components/Input";
import Select from "components/Select";
import React, { ChangeEvent, HTMLAttributes } from "react";
import classnames from "classnames";
import {
  generatePagingOptions,
  getDescription,
  useFigureSmallScreen,
} from "../utils";
import {
  BasePaginationProps,
  ExternalControlProps,
  InternalControlProps,
} from "../TablePagination";

export enum Label {
  NEXT_PAGE = "Next page",
  PREVIOUS_PAGE = "Previous page",
  PAGE_NUMBER = "Page number",
}

export type AllProps = BasePaginationProps &
  InternalControlProps &
  ExternalControlProps;

export type Props = Omit<
  AllProps,
  | "currentPage"
  | "data"
  | "dataForwardProp"
  | "externallyControlled"
  | "onPageChange"
  | "position"
  | "totalItems"
> & {
  currentPage?: AllProps["currentPage"];
  displayDescription?: boolean;
  onInputPageChange?: (page: number) => void;
  nextButtonProps?: Partial<ButtonProps>;
  onNextPage?: (page: number) => void;
  onPageChange?: AllProps["onPageChange"];
  onPreviousPage?: (page: number) => void;
  previousButtonProps?: Partial<ButtonProps>;
  totalItems?: AllProps["totalItems"];
  visibleCount?: number;
  showPageInput?: boolean;
} & HTMLAttributes<HTMLDivElement>;

const TablePaginationControls = ({
  className,
  currentPage,
  description,
  displayDescription = true,
  onInputPageChange,
  itemName,
  nextButtonProps,
  onNextPage,
  onPageChange,
  onPageSizeChange,
  onPreviousPage,
  pageLimits,
  pageSize,
  previousButtonProps,
  showPageInput = true,
  totalItems,
  visibleCount,
  ...divProps
}: Props): JSX.Element => {
  const isSmallScreen = useFigureSmallScreen();

  const totalPages = totalItems ? Math.ceil(totalItems / pageSize) : null;
  const descriptionDisplay = getDescription({
    description,
    visibleCount,
    isSmallScreen,
    totalItems,
    itemName,
  });

  const handleDecrementPage = (currentPage: number) => {
    if (currentPage > 1) {
      onPageChange?.(currentPage - 1);
    }
    onPreviousPage?.(typeof currentPage === "number" ? currentPage - 1 : null);
  };

  const handleIncrementPage = (currentPage: number, totalPages: number) => {
    if (currentPage < totalPages) {
      onPageChange?.(currentPage + 1);
    }
    onNextPage?.(typeof currentPage === "number" ? currentPage + 1 : null);
  };

  const handleInputPageChange = (e: ChangeEvent<HTMLInputElement>) => {
    const newPage = Math.min(totalPages, Math.max(1, parseInt(e.target.value)));
    onPageChange?.(newPage);
    onInputPageChange?.(Number(e.target.value));
  };

  const handlePageSizeChange = (e: ChangeEvent<HTMLSelectElement>) => {
    onPageSizeChange(parseInt(e.target.value));
  };

  return (
    <div
      className={classnames("pagination", className)}
      {...divProps}
      role="navigation"
    >
      <div className="description" id="pagination-description">
        {displayDescription ? descriptionDisplay : null}
      </div>
      <Button
        aria-label={Label.PREVIOUS_PAGE}
        className="back"
        appearance="base"
        hasIcon
        disabled={currentPage === 1}
        onClick={() => handleDecrementPage(currentPage)}
        {...previousButtonProps}
      >
        {previousButtonProps?.children ? (
          previousButtonProps.children
        ) : (
          <Icon name="chevron-down" />
        )}
      </Button>
      {showPageInput ? (
        <>
          <Input
            id="paginationPageInput"
            label={Label.PAGE_NUMBER}
            labelClassName="u-off-screen"
            className="u-no-margin--bottom pagination-input"
            onChange={handleInputPageChange}
            value={currentPage}
            type="number"
          />{" "}
        </>
      ) : null}
      {typeof totalPages === "number" ? `of ${totalPages}` : null}
      <Button
        aria-label={Label.NEXT_PAGE}
        className="next"
        appearance="base"
        hasIcon
        disabled={currentPage === totalPages}
        onClick={() => handleIncrementPage(currentPage, totalPages)}
        {...nextButtonProps}
      >
        <Icon name="chevron-down" />
      </Button>
      <Select
        className="u-no-margin--bottom"
        label="Items per page"
        labelClassName="u-off-screen"
        id="itemsPerPage"
        options={generatePagingOptions(pageLimits)}
        onChange={handlePageSizeChange}
        value={pageSize}
      />
    </div>
  );
};

export default TablePaginationControls;
