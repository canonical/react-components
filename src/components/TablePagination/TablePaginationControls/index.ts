export { default } from "./TablePaginationControls";
export type { Props as TablePaginationControlsProps } from "./TablePaginationControls";
