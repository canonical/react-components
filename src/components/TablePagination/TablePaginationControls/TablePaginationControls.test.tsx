import React from "react";
import { render, screen } from "@testing-library/react";

import TablePaginationControls from "./TablePaginationControls";
import userEvent from "@testing-library/user-event";
import { Label } from "./TablePaginationControls";

describe("<TablePaginationControls />", () => {
  // snapshot tests
  it("renders table pagination controls and matches the snapshot", () => {
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={0}
        pageSize={20}
        onPageChange={jest.fn()}
        onPageSizeChange={jest.fn()}
      />,
    );

    expect(screen.getAllByRole("button")).toMatchSnapshot();
    expect(screen.getByRole("spinbutton")).toMatchSnapshot();
  });

  it("can go to the next page", async () => {
    const onPageChange = jest.fn();
    const onNextPage = jest.fn();
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={onPageChange}
        onNextPage={onNextPage}
        onPageSizeChange={jest.fn()}
      />,
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.NEXT_PAGE }),
    );
    expect(onPageChange).toHaveBeenCalledWith(3);
    expect(onNextPage).toHaveBeenCalledWith(3);
  });

  it("can go to the previous page", async () => {
    const onPageChange = jest.fn();
    const onPreviousPage = jest.fn();
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={onPageChange}
        onPreviousPage={onPreviousPage}
        onPageSizeChange={jest.fn()}
      />,
    );
    await userEvent.click(
      screen.getByRole("button", { name: Label.PREVIOUS_PAGE }),
    );
    expect(onPageChange).toHaveBeenCalledWith(1);
    expect(onPreviousPage).toHaveBeenCalledWith(1);
  });

  it("can set the page using the input", async () => {
    const onInputPageChange = jest.fn();
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={jest.fn()}
        onPreviousPage={jest.fn()}
        onPageSizeChange={jest.fn()}
        onInputPageChange={onInputPageChange}
      />,
    );
    await userEvent.type(
      screen.getByRole("spinbutton", { name: Label.PAGE_NUMBER }),
      "1",
    );
    expect(onInputPageChange).toHaveBeenCalledWith(21);
  });

  it("can hide the page input", async () => {
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={jest.fn()}
        onPreviousPage={jest.fn()}
        onPageSizeChange={jest.fn()}
        showPageInput={false}
      />,
    );
    expect(
      screen.queryByRole("spinbutton", { name: Label.PAGE_NUMBER }),
    ).not.toBeInTheDocument();
  });

  it("can display the description", async () => {
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={jest.fn()}
        onPreviousPage={jest.fn()}
        onPageSizeChange={jest.fn()}
        showPageInput={false}
      />,
    );
    expect(document.querySelector(".description")?.textContent).toBe(
      "Showing 5 out of 100 items",
    );
  });

  it("can hide the description", async () => {
    render(
      <TablePaginationControls
        visibleCount={5}
        itemName="item"
        pageLimits={[20, 50, 100]}
        totalItems={100}
        currentPage={2}
        pageSize={20}
        onPageChange={jest.fn()}
        onPreviousPage={jest.fn()}
        onPageSizeChange={jest.fn()}
        displayDescription={false}
      />,
    );
    expect(
      document.querySelector(".description")?.textContent,
    ).not.toBeUndefined();
  });
});
