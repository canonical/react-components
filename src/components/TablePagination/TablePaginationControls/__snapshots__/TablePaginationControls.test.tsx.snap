// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TablePaginationControls /> renders table pagination controls and matches the snapshot 1`] = `
[
  <button
    aria-label="Previous page"
    class="p-button--base has-icon back"
  >
    <i
      class="p-icon--chevron-down"
    />
  </button>,
  <button
    aria-label="Next page"
    class="p-button--base has-icon next"
  >
    <i
      class="p-icon--chevron-down"
    />
  </button>,
]
`;

exports[`<TablePaginationControls /> renders table pagination controls and matches the snapshot 2`] = `
<input
  aria-invalid="false"
  class="p-form-validation__input u-no-margin--bottom pagination-input"
  id="paginationPageInput"
  type="number"
  value="0"
/>
`;
